api/server/router/container/exec.go

```diff
@@ -92,11 +92,19 @@ func (s *containerRouter) postContainerExecStart(ctx context.Context, w http.Res
 		defer httputils.CloseStreams(inStream, outStream)
 
 		if _, ok := r.Header["Upgrade"]; ok {
-			fmt.Fprintf(outStream, "HTTP/1.1 101 UPGRADED\r\nContent-Type: application/vnd.docker.raw-stream\r\nConnection: Upgrade\r\nUpgrade: tcp\r\n\r\n")
+			fmt.Fprintf(outStream, "HTTP/1.1 101 UPGRADED\r\nContent-Type: application/vnd.docker.raw-stream\r\nConnection: Upgrade\r\nUpgrade: tcp\r\n")
 		} else {
-			fmt.Fprintf(outStream, "HTTP/1.1 200 OK\r\nContent-Type: application/vnd.docker.raw-stream\r\n\r\n")
+			fmt.Fprintf(outStream, "HTTP/1.1 200 OK\r\nContent-Type: application/vnd.docker.raw-stream\r\n")
 		}
 
+		// copy headers that were removed as part of hijack
+		for k, v := range w.Header() {
+			for _, i := range v {
+				fmt.Fprintf(outStream, "%s: %s\r\n", k, i)
```

Comment:
If the headers or their values contain newlines, that will lead to header injection. Any situation where user-controlled data ends up inside a header would pose a risk. There may be some other place in the code where header data is sanitized, or we may currently assume that user-controlled data never ends up headers. Regardless, it is good practice to sanitize headers at the point where they are output to the stream.